disco_double:

  # 话题
  pointCloudTopic: "points_raw"               # 点云数据
  imuTopic: "imu_raw"                         # IMU数据
  odomTopic: "odometry/imu"                   # IMU预积分里程计，与IMU频率相同
  gpsTopic: "odometry/gpsz"                   # 来自navsat的GPS里程计话题，参见module_navsat.launch文件

  # 坐标系
  lidarFrame: "base_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # GPS设置
  useImuHeadingInitialization: false           # 如果使用GPS数据，设置为"true"
  useGpsElevation: false                      # 如果GPS高程数据不好，设置为"false"
  gpsCovThreshold: 2.0                        # 平方米，使用GPS数据的阈值
  poseCovThreshold: 25.0                      # 平方米，使用GPS数据的阈值
  
  # 导出设置
  savePCD: false                              # https://github.com/TixiaoShan/LIO-SAM/issues/3
  savePCDDirectory: "/Downloads/LOAM/"        # 在您的主文件夹中，以"/"开始和结束。警告：代码会删除"LOAM"文件夹然后重新创建它。具体实现见"mapOptimization"

  # 传感器设置
  sensor: velodyne                            # 激光雷达传感器类型，'velodyne'或'ouster'或'livox'
  N_SCAN: 32                                  # 激光雷达通道数量（例如，Velodyne/Ouster: 16, 32, 64, 128, Livox Horizon: 6）
  Horizon_SCAN: 1800                          # 激光雷达水平分辨率（Velodyne:1800, Ouster:512,1024,2048, Livox Horizon: 4000）
  downsampleRate: 1                           # 默认：1。如果点太多，可以降采样数据。例如，16 = 64 / 4，16 = 16 / 1
  lidarMinRange: 1.0                          # 默认：1.0，使用的最小激光雷达范围
  lidarMaxRange: 1000.0                       # 默认：1000.0，使用的最大激光雷达范围

  # IMU设置
  imuAccNoise: 3.9939570888238808e-03
  imuGyrNoise: 1.5636343949698187e-03
  imuAccBiasN: 6.4356659353532566e-05
  imuGyrBiasN: 3.5640318696367613e-05
  imuGravity: 9.80511
  imuRPYWeight: 0.01

  # 外参：T_lb（激光雷达 -> IMU）
  extrinsicTrans: [0.03499839, -0.00199992, 1.22998917]
  #extrinsicRot: [-1, 0, 0,
  #                0, 1, 0,
  #                0, 0, -1]
  #extrinsicRPY: [0, -1, 0,
  #               1, 0, 0,
  #               0, 0, 1]
  extrinsicRot: [0, -1, 0,
                   0, 1, 0,
                   0, 0, 1]
  extrinsicRPY: [1, 0, 0,
                   0, 1, 0,
                   0, 0, 1]
  
  # LOAM特征阈值
  edgeThreshold: 1.0
  surfThreshold: 0.1
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100

  # 体素滤波参数
  odometrySurfLeafSize: 0.4                     # 默认：0.4 - 室外，0.2 - 室内
  mappingCornerLeafSize: 0.2                    # 默认：0.2 - 室外，0.1 - 室内
  mappingSurfLeafSize: 0.4                      # 默认：0.4 - 室外，0.2 - 室内

  # 机器人运动约束（适用于2D机器人）
  z_tollerance: 1000                            # 米
  rotation_tollerance: 1000                     # 弧度

  # CPU参数
  numberOfCores: 4                              # 用于地图优化的核心数
  mappingProcessInterval: 0.15                  # 秒，调节地图构建频率

  # 周围地图
  surroundingkeyframeAddingDistThreshold: 1.0   # 米，调节关键帧添加阈值
  surroundingkeyframeAddingAngleThreshold: 0.2  # 弧度，调节关键帧添加阈值
  surroundingKeyframeDensity: 2.0               # 米，下采样周围关键帧位姿   
  surroundingKeyframeSearchRadius: 50.0         # 米，在n米范围内进行扫描到地图的优化（当回环检测禁用时）

  # 回环检测
  loopClosureEnableFlag: true
  loopClosureFrequency: 1.0                     # 赫兹，调节回环约束添加频率
  surroundingKeyframeSize: 50                   # 子地图大小（当回环检测启用时）
  historyKeyframeSearchRadius: 15.0             # 米，距离当前位姿n米内的关键帧将被考虑用于回环检测
  historyKeyframeSearchTimeDiff: 30.0           # 秒，n秒以前的关键帧将被考虑用于回环检测
  historyKeyframeSearchNum: 25                  # 将融合到子地图中用于回环检测的历史关键帧数量
  historyKeyframeFitnessScore: 0.3              # ICP阈值，越小表示对齐越好

  # 可视化
  globalMapVisualizationSearchRadius: 1000.0    # 米，全局地图可视化半径
  globalMapVisualizationPoseDensity: 10.0       # 米，全局地图可视化关键帧密度
  globalMapVisualizationLeafSize: 1.0           # 米，全局地图可视化点云密度




# Navsat（将GPS坐标转换为笛卡尔坐标）
navsat:
  frequency: 50
  wait_for_datum: false
  delay: 0.0
  magnetic_declination_radians: 0
  yaw_offset: 0
  zero_altitude: true
  broadcast_utm_transform: false
  broadcast_utm_transform_as_parent_frame: false
  publish_filtered_gps: false

# 用于Navsat的EKF
ekf_gps:
  publish_tf: false
  map_frame: map
  odom_frame: odom
  base_link_frame: base_link
  world_frame: odom

  frequency: 50
  two_d_mode: false
  sensor_timeout: 0.01
  # -------------------------------------
  # 外部IMU:
  # -------------------------------------
  imu0: imu_correct
  # 确保输入与ROS REP105对齐。"imu_correct"是我手动转换的。EKF也可以使用IMU和base_link之间的tf来转换数据
  imu0_config: [false, false, false,
                true,  true,  true,
                false, false, false,
                false, false, true,
                true,  true,  true]
  imu0_differential: false
  imu0_queue_size: 50 
  imu0_remove_gravitational_acceleration: true
  # -------------------------------------
  # 里程计（来自Navsat）:
  # -------------------------------------
  odom0: odometry/gps
  odom0_config: [true,  true,  true,
                 false, false, false,
                 false, false, false,
                 false, false, false,
                 false, false, false]
  odom0_differential: false
  odom0_queue_size: 10

  #                            x     y     z     r     p     y   x_dot  y_dot  z_dot  r_dot p_dot y_dot x_ddot y_ddot z_ddot
  process_noise_covariance: [  1.0,  0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    1.0,  0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    10.0, 0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0.03, 0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0.03, 0,    0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0.1,  0,     0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0.25,  0,     0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0.25,  0,     0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0.04,  0,    0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0.01, 0,    0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0.01, 0,    0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0.5,  0,    0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0.01, 0,      0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0.01,   0,
                               0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,    0,    0,      0.015]
